source/uwp/Renderer/lib/XamlBuilder.cpp

```diff
@@ -1126,6 +1127,253 @@ namespace AdaptiveCards { namespace Rendering { namespace Uwp
+
+        ComPtr<ABI::Windows::Data::Xml::Dom::IXmlDocumentIO> xmlDocumentIO;
+        RETURN_IF_FAILED(xmlDocument.As(&xmlDocumentIO));
+        
+        HRESULT hr = xmlDocumentIO->LoadXml(htmlHString.Get());
+        if (FAILED(hr))
+        {
+            RETURN_IF_FAILED(textBlock->put_Text(textIn));
+        }
+        else
+        {
+            ComPtr<IVector<ABI::Windows::UI::Xaml::Documents::Inline*>> inlines;
+            RETURN_IF_FAILED(textBlock->get_Inlines(inlines.GetAddressOf()));
+
+            ComPtr<ABI::Windows::Data::Xml::Dom::IXmlNode> xmlDocumentAsNode;
+            RETURN_IF_FAILED(xmlDocument.As(&xmlDocumentAsNode));
+
+            ComPtr<ABI::Windows::Data::Xml::Dom::IXmlNode> pNode;
+            RETURN_IF_FAILED(xmlDocumentAsNode->get_FirstChild(&pNode));
+
+            HString nodeName;
+            RETURN_IF_FAILED(pNode->get_NodeName(nodeName.GetAddressOf()));
+
+            INT32 isOrderedListResult;
+            RETURN_IF_FAILED(WindowsCompareStringOrdinal(nodeName.Get(), HStringReference(L"ol").Get(), &isOrderedListResult));
+
+            INT32 isUnorderedListResult;
+            RETURN_IF_FAILED(WindowsCompareStringOrdinal(nodeName.Get(), HStringReference(L"ul").Get(), &isUnorderedListResult));
+
+            if ((isOrderedListResult == 0) || (isUnorderedListResult == 0))
+            {
+                RETURN_IF_FAILED(AddListInlines(pNode.Get(), (isOrderedListResult == 0), inlines.Get()));
+            }
+            else
+            {
+                ComPtr<ABI::Windows::Data::Xml::Dom::IXmlNode> childNode;
+                RETURN_IF_FAILED(pNode->get_FirstChild(&childNode));
+
+                while (childNode != nullptr)
+                {
+                    HString nodeName;
+                    RETURN_IF_FAILED(childNode->get_NodeName(nodeName.GetAddressOf()));
+
+                    INT32 isLinkResult;
+                    RETURN_IF_FAILED(WindowsCompareStringOrdinal(nodeName.Get(), HStringReference(L"a").Get(), &isLinkResult));
+
+                    INT32 isBoldResult;
+                    RETURN_IF_FAILED(WindowsCompareStringOrdinal(nodeName.Get(), HStringReference(L"strong").Get(), &isBoldResult));
+
+                    INT32 isItalicResult;
+                    RETURN_IF_FAILED(WindowsCompareStringOrdinal(nodeName.Get(), HStringReference(L"em").Get(), &isItalicResult));
+
+                    if (isLinkResult == 0)
+                    {
+                        RETURN_IF_FAILED(AddLinkInline(childNode.Get(), inlines.Get()));
+                    }
+                    else
+                    {
+                        RETURN_IF_FAILED(AddTextInline(renderContext, childNode.Get(), (isBoldResult == 0), (isItalicResult == 0), inlines.Get()));
+                    }
+
+                    ComPtr<ABI::Windows::Data::Xml::Dom::IXmlNode> nextChildNode;
+                    RETURN_IF_FAILED(childNode->get_NextSibling(&nextChildNode));
```

Comment:
i don't know how exactly xml parses html... what happens if html tag has child tag for example <link tag "<- i just made it up" > <bold> "hello wordl" </bold> </link tag> will get_NextSibling returns correct node?

Reply:
I think your concern is fixed by my rework. Let me know if you still have concerns on the new iteration.

---
In reply to: [167377042](https://github.com/Microsoft/AdaptiveCards/pull/1113#discussion_r167377042) [](ancestors = 167377042)